compiler/rustc_mir/src/const_eval/error.rs

```diff
@@ -22,12 +22,15 @@ pub enum ConstEvalErrKind {
     Panic { msg: Symbol, line: u32, col: u32, file: Symbol },
 }
 
-// The errors become `MachineStop` with plain strings when being raised.
+// Non-panic errors become `MachineStop` with plain strings when being raised.
 // `ConstEvalErr` (in `librustc_middle/mir/interpret/error.rs`) knows to
 // handle these.
 impl<'tcx> Into<InterpErrorInfo<'tcx>> for ConstEvalErrKind {
     fn into(self) -> InterpErrorInfo<'tcx> {
-        err_machine_stop!(self.to_string()).into()
+        match self {
+            ConstEvalErrKind::Panic { msg, .. } => InterpError::Panic(msg).into(),
```

Comment:
I think we can mirror [what miri does](https://github.com/rust-lang/miri/blob/5fdea5be99eb44ea2b3adebb1f8a3c1c244ffd7e/src/diagnostics.rs#L13) here. Instead of having an `impl MachineStopType` for `String`, we create an

```rust
enum TerminationInfo {
    Generic(String),
    Panic(String),
}
```

Then you can downcast to `TerminationInfo` below and handle it just there.